pkg/filters/opafilter/opafilter.go

```diff
@@ -0,0 +1,241 @@
+	regoQuery             *rego.PreparedEvalQuery
+}
+
+type Spec struct {
+	filters.BaseSpec `yaml:",inline"`
+	DefaultStatus    int    `yaml:"defaultStatus"`
+	IncludedHeaders  string `yaml:"includedHeaders"`
+	ReadBody         bool   `yaml:"readBody"`
+	Policy           string `yaml:"policy"`
+}
+
+func init() {
+	filters.Register(kind)
+}
+
+func (o *OPAFilter) Name() string {
+	return o.spec.Name()
+}
+
+func (o *OPAFilter) Spec() filters.Spec {
+	return o.spec
+}
+
+func (o *OPAFilter) Kind() *filters.Kind {
+	return kind
+}
+
+func (o *OPAFilter) Init() {
+	o.includedHeadersParsed = strings.Split(o.spec.IncludedHeaders, ",")
+	if o.spec.DefaultStatus == 0 {
+		o.spec.DefaultStatus = 403
+	}
+	n := 0
+	for i := range o.includedHeadersParsed {
+		scrubbed := strings.ReplaceAll(o.includedHeadersParsed[i], " ", "")
+		if scrubbed != "" {
+			o.includedHeadersParsed[n] = textproto.CanonicalMIMEHeaderKey(scrubbed)
+			n++
+		}
+	}
+	o.includedHeadersParsed = o.includedHeadersParsed[:n]
+	ctx, cancelFunc := stdctx.WithTimeout(stdctx.Background(), 5*time.Second)
+	query, err := rego.New(
+		rego.Query("result = data.http.allow"),
+		rego.Module("inline.rego.policy", o.spec.Policy),
+	).PrepareForEval(ctx)
+	defer cancelFunc()
+	if err != nil {
+		logger.Errorf("create PrepareForEval rego query error: %s", err)
+	} else {
+		o.regoQuery = &query
+	}
+}
+
+func (o *OPAFilter) Inherit(previousGeneration filters.Filter) {
+	o.Init()
+	previousGeneration.Close()
+}
+
+func (o *OPAFilter) Handle(ctx *context.Context) (result string) {
+	req := ctx.GetInputRequest().(*httpprot.Request)
+	var rw *httpprot.Response
+	if rw, _ = ctx.GetOutputResponse().(*httpprot.Response); rw == nil {
+		rw, _ = httpprot.NewResponse(nil)
+		ctx.SetOutputResponse(rw)
+	}
+	return o.evalRequest(req, rw)
+}
+
+func (o *OPAFilter) Status() interface{} {
+	return nil
+}
+
+func (o *OPAFilter) Close() {
+}
+
+func (o *OPAFilter) evalRequest(r *httpprot.Request, w *httpprot.Response) string {
+	headers := map[string]string{}
+
+	for key, value := range r.HTTPHeader() {
+		if len(value) > 0 && stringtool.StrInSlice(key, o.includedHeadersParsed) {
+			headers[key] = strings.Join(value, ", ")
+		}
+	}
+
+	var body string
+	if o.spec.ReadBody {
+		buf, _ := io.ReadAll(r.Body)
+		body = string(buf)
+		// Put the body back in the request
+		r.Body = io.NopCloser(bytes.NewBuffer(buf))
+	}
+	pathParts := strings.Split(strings.Trim(r.Path(), "/"), "/")
+	input := map[string]any{
+		"request": map[string]any{
+			"method":     r.Std().Method,
+			"path":       r.Std().URL.Path,
+			"path_parts": pathParts,
+			"raw_query":  r.Std().URL.RawQuery,
+			"query":      map[string][]string(r.Std().URL.Query()),
+			"headers":    headers,
+			"scheme":     r.Scheme(),
+			"realIP":     r.RealIP(),
+			"body":       body,
+		},
+	}
+	results, err := o.regoQuery.Eval(r.Context(), rego.EvalInput(input))
+	if err != nil {
+		return o.opaError(w, err)
+	}
+
+	if len(results) == 0 {
```

Comment:
do you think it is nessesary to check if len(results) > 1 and then report error?